### PR TITLE
CMake updates

### PR DESCRIPTION
**Pull request by Optiligence**
todos: 
- [ ] make the legacy target (`quazip`) opt-in
- [ ] move the private headers into a subdirectory
- [ ] some explanatory comments/reasoning
- [ ] readme for build/consumption
- [ ] pkg-config?
- [ ] maybe split it up into more handy commits again
- [ ] qmake?

please try it out: 
`mkdir build && cd build && cmake -G Ninja -DCMAKE_INSTALL_PREFIX=install ..`
```
cmake -DBUILD_SHARED_LIBS=0 -DCMAKE_BUILD_TYPE=DEBUG . && ninja check && ninja install && 
cmake -DBUILD_SHARED_LIBS=1 -DCMAKE_BUILD_TYPE=DEBUG . && ninja check && ninja install && 
cmake -DBUILD_SHARED_LIBS=0 -DCMAKE_BUILD_TYPE=RELEASE . && ninja check && ninja install && 
cmake -DBUILD_SHARED_LIBS=1 -DCMAKE_BUILD_TYPE=RELEASE . && ninja check && ninja install
```
`find install/`
```
install/
install/bin
install/bin/libquazip.dll
install/bin/libquazip1-qt5.dll
install/bin/libquazip1-qt5d.dll
install/bin/libquazipd.dll
install/include
install/include/quazip1-qt5
install/include/quazip1-qt5/quazip
install/include/quazip1-qt5/quazip/ioapi.h
install/include/quazip1-qt5/quazip/JlCompress.h
install/include/quazip1-qt5/quazip/minizip_crypt.h
install/include/quazip1-qt5/quazip/quaadler32.h
install/include/quazip1-qt5/quazip/quachecksum32.h
install/include/quazip1-qt5/quazip/quacrc32.h
install/include/quazip1-qt5/quazip/quagzipfile.h
install/include/quazip1-qt5/quazip/quaziodevice.h
install/include/quazip1-qt5/quazip/quazip.h
install/include/quazip1-qt5/quazip/quazipdir.h
install/include/quazip1-qt5/quazip/quazipfile.h
install/include/quazip1-qt5/quazip/quazipfileinfo.h
install/include/quazip1-qt5/quazip/quazipnewinfo.h
install/include/quazip1-qt5/quazip/quazip_global.h
install/include/quazip1-qt5/quazip/unzip.h
install/include/quazip1-qt5/quazip/zip.h
install/lib
install/lib/cmake
install/lib/cmake/QuaZip
install/lib/cmake/QuaZip/QuaZipConfig.cmake
install/lib/cmake/QuaZip/QuaZipConfigVersion.cmake
install/lib/cmake/QuaZip/QuaZipSharedQt5Targets-debug.cmake
install/lib/cmake/QuaZip/QuaZipSharedQt5Targets-release.cmake
install/lib/cmake/QuaZip/QuaZipSharedQt5Targets.cmake
install/lib/cmake/QuaZip/QuaZipStaticQt5Targets-debug.cmake
install/lib/cmake/QuaZip/QuaZipStaticQt5Targets-release.cmake
install/lib/cmake/QuaZip/QuaZipStaticQt5Targets.cmake
install/lib/libquazip.a
install/lib/libquazip.dll.a
install/lib/libquazip1-qt5.a
install/lib/libquazip1-qt5.dll.a
install/lib/libquazip1-qt5d.a
install/lib/libquazip1-qt5d.dll.a
install/lib/libquazipd.a
install/lib/libquazipd.dll.a
install/lib/pkgconfig
install/lib/pkgconfig/quazip.pc
```
can be used with
```
find_package(QuaZip REQUIRED COMPONENTS Qt5)
target_link_libraries(${PROJECT_NAME} QuaZip::Qt5)
```
@stachenov i think it would be best if you go through everything in detail and ask me to elaborate on any specifics

### CODE REVIEW

**Comment by stachenov** [comment]
Wait a moment. This should be split not into different *commits*, but into different *PRs* (by proper branching). What does that segfault has to do with CMake or test warnings? Then these can be discussed separately too.

**Comment by Optiligence** [comment]
please ignore those, they are just here so i could build it and see properly (without the warning spam)

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.5)
+cmake_minimum_required(VERSION 3.14) # default install paths
 project(QuaZip)
 
 # CMP0042: Explicitly acknowledge MACOSX_RPATH
```
Comment (Highlife1911) [inline CMakeLists.txt:4]: consider https://cmake.org/cmake/help/latest/command/cmake_policy.html#setting-policies-by-cmake-version to set everything to new. Usually the newer policies are better:
`cmake_policy(VERSION 3.14.0)`

File: CMakeLists.txt [hunk 2]
```diff
@@ -9,54 +9,49 @@ cmake_policy(SET CMP0042 NEW)
...
+enable_testing()
+set(CMAKE_AUTOMOC ON)
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CMAKE_DEBUG_POSTFIX d)
+
+set(QUAZIP_LIB_FILE_NAME quazip)
+if(WIN32)
+	set(QUAZIP_LIB_FILE_NAME ${QUAZIP_LIB_FILE_NAME}${QUAZIP_LIB_SOVERSION})
+endif()
+if(NOT BUILD_WITH_QT4)
```
Comment (Highlife1911) [inline CMakeLists.txt:28]: reverse if/else branch and remove the `NOT`

File: CMakeLists.txt [hunk 3]
```diff
@@ -9,54 +9,49 @@ cmake_policy(SET CMP0042 NEW)
...
+	set(QUAZIP_QTCORE_LIBRARIES Qt4::QtCore)
+	set(QUAZIP_LIB_FILE_NAME ${QUAZIP_LIB_FILE_NAME}-qt4)
+	set(QUAZIP_LIB_TARGET_NAME Qt4)
 endif()
 
 find_package(ZLIB REQUIRED)
 
-set(QUAZIP_LIB_TARGET_NAME ${QUAZIP_LIB_VERSION_PREFIX}Quazip CACHE
-    INTERNAL "Target name of libquazip" FORCE)
+set(QUAZIP_LIB_NAME ${QUAZIP_LIB_FILE_NAME} CACHE INTERNAL "Target name of libquazip" FORCE)
```
Comment (Highlife1911) [inline CMakeLists.txt:43]: I don't understand why this must be `CACHE INTERNAL`

Comment (Highlife1911) [reply CMakeLists.txt:43]: Is this for debugging?

File: CMakeLists.txt [hunk 4]
```diff
@@ -9,54 +9,49 @@ cmake_policy(SET CMP0042 NEW)
...
-    configure_file(quazip.pc.cmakein quazip.pc @ONLY)
-    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/quazip.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
-endif()
-
-install(FILES QuaZipConfig.cmake
-        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/QuaZip${QUAZIP_LIB_VERSION_SUFFIX}
-        RENAME QuaZip${QUAZIP_LIB_VERSION_SUFFIX}Config.cmake)
+add_library(QuaZip::QuaZip ALIAS ${QUAZIP_LIB_TARGET_NAME})
+add_subdirectory(qztest EXCLUDE_FROM_ALL)
+add_test(NAME qztest_test
```
Comment (Highlife1911) [inline CMakeLists.txt:49]: It hint this test stuff should be part of the `CMakeLists.txt` from `qztest`

File: qztest/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+project(qztest)
+
+find_package(Qt5 REQUIRED COMPONENTS Network Test)
```
Comment (Highlife1911) [inline qztest/CMakeLists.txt:3]: What about tests for Qt4? `BUILD_WITH_QT4` implies that we work with Qt4 but tests are always using Qt5

File: CMakeLists.txt [hunk 6]
```diff
@@ -9,54 +9,49 @@ cmake_policy(SET CMP0042 NEW)
 set(QUAZIP_LIB_VERSION 1.0.0)
 set(QUAZIP_LIB_SOVERSION 1)
 
+option(BUILD_SHARED_LIBS "" ON)
```
Comment (Highlife1911) [inline CMakeLists.txt:12]: this looks unused

**Comment by stachenov** [comment]
Oops... You're commenting on a PR that is already merged with a lot of changes added on top of it. Many of the issues you mention are already fixed there. Please have a look at `cmake` branch and comment at #82.

**Pull request merged**